drivers/crypto/crypto_rts5912_sha.c

```diff
@@ -0,0 +1,310 @@
+			return -EIO;
+		}
+
+		if (sha2dma_regs->interrupt_status & SHA2DMA_INTSTS_BUS_COMPLETE_Msk) {
+			return -EIO;
+		}
+
+		sha2dma_regs->clear_transfer = SHA2DMA_INTCLR_CLRTFR_Msk;
+		idx += RTS5912_SHA2DMA_MAXIMUM_BLOCK_NUM << RTS5912_SHA2DMA_BLOCK_SHIFT;
+
+		if (blk_size <= RTS5912_SHA2DMA_MAXIMUM_BLOCK_NUM) {
+			break;
+		}
+	}
+	k_mutex_lock(&crypto_rts5912_in_use, K_FOREVER);
+	for (uint32_t i = 0; i < 8; i++) {
+		rts5912_sha256_ctx->state[i] = sha2_regs->digest[i << 1];
+	}
+	k_mutex_unlock(&crypto_rts5912_in_use);
+	return 0;
+}
+
+static int rts5912_sha256_update(const struct device *dev, uint8_t *input, size_t len)
+{
+	struct rts5912_sha256_context *rts5912_sha256_ctx = dev->data;
+	uint32_t remain, fill, blk_size = 0, ret_val = 0;
+
+	remain = rts5912_sha256_ctx->total[0] & (RTS5912_SHA2DMA_BLOCK_SIZE - 1);
+	fill = RTS5912_SHA2DMA_BLOCK_SIZE - remain;
+
+	k_mutex_lock(&crypto_rts5912_in_use, K_FOREVER);
+	rts5912_sha256_ctx->total[0] += len;
+	if (rts5912_sha256_ctx->total[0] < len) {
+		rts5912_sha256_ctx->total[1]++;
+	}
+	k_mutex_unlock(&crypto_rts5912_in_use);
+
+	if ((len >= fill) && (remain != 0)) {
+		k_mutex_lock(&crypto_rts5912_in_use, K_FOREVER);
+		memcpy((void *)(&(rts5912_sha256_ctx->buffer[remain])), input, fill);
+		k_mutex_unlock(&crypto_rts5912_in_use);
+		ret_val = rts5912_sha256_process(dev, (&(rts5912_sha256_ctx->buffer[0])), 0x1);
+		if (ret_val != 0) {
+			return ret_val;
+		}
+
+		input += fill;
+		len -= fill;
+		remain = 0;
+	}
+
+	while (len >= RTS5912_SHA2DMA_BLOCK_SIZE) {
+		blk_size = len >> RTS5912_SHA2DMA_BLOCK_SHIFT;
+		ret_val = rts5912_sha256_process(dev, input, blk_size);
+		if (ret_val != 0) {
+			return ret_val;
+		}
+		input += (len & ~(RTS5912_SHA2DMA_BLOCK_SIZE - 1));
+		len &= (RTS5912_SHA2DMA_BLOCK_SIZE - 1);
+	}
+
+	k_mutex_lock(&crypto_rts5912_in_use, K_FOREVER);
+	if (len > 0) {
+		memcpy((void *)(&(rts5912_sha256_ctx->buffer[remain])), input, len);
+	}
+	k_mutex_unlock(&crypto_rts5912_in_use);
+
+	return ret_val;
+}
+
+static int rts5912_sha256_finish(const struct device *dev, uint8_t *output)
+{
+	struct rts5912_sha256_context *rts5912_sha256_ctx = dev->data;
+	uint32_t remain, be_high, be_low, ret_val = 0;
+
+	k_mutex_lock(&crypto_rts5912_in_use, K_FOREVER);
+	remain = rts5912_sha256_ctx->total[0] & (RTS5912_SHA2DMA_BLOCK_SIZE - 1);
+	rts5912_sha256_ctx->buffer[remain++] = 0x80U;
+
+	if (remain <= RTS5912_SHA2_BLOCK_EXTEND_CHECK) {
+		memset((void *)(&(rts5912_sha256_ctx->buffer[remain])), 0x0,
+		       RTS5912_SHA2_BLOCK_EXTEND_CHECK - remain);
+	} else {
+		memset((void *)(&(rts5912_sha256_ctx->buffer[remain])), 0x0,
+		       RTS5912_SHA2DMA_BLOCK_SIZE - remain);
+		k_mutex_unlock(&crypto_rts5912_in_use);
+		ret_val = rts5912_sha256_process(dev, (uint8_t *)(&(rts5912_sha256_ctx->buffer[0])),
+						 0x1);
+		k_mutex_lock(&crypto_rts5912_in_use, K_FOREVER);
+		if (ret_val != 0) {
+			k_mutex_unlock(&crypto_rts5912_in_use);
+			return ret_val;
+		}
```

Comment:
Check the return value before locking the mutex.    No need to lock the mutex just to release it when you know there is an error.